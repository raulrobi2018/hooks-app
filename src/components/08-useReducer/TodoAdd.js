import React from "react";
import {useForm} from "../../hooks/useForm";

export const TodoAdd = ({addTodo}) => {
    // Utilizo mi custom hook definido en el archivo useForm.js
    // Extraigo la descripción del estado devuelto y la función que deseo manejar
    const [{description}, {handleInputChange, reset}] = useForm({
        description: ""
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length <= 1) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        addTodo(newTodo);
        reset();
    };

    return (
        <form className="mb-3">
            <div className="row">
                <div className="col-8">
                    <div className="input-group mb-3">
                        <span className="input-group-text" id="basic-addon1">
                            Description
                        </span>
                        <input
                            type="text"
                            className="form-control"
                            aria-label="Description"
                            aria-describedby="basic-addon1"
                            name="description"
                            autoComplete="off"
                            onChange={handleInputChange}
                            value={description}
                        />
                    </div>
                </div>
                <div className="col-2">
                    <button
                        className="btn btn-success btn-block"
                        type="submit"
                        onClick={handleSubmit}
                    >
                        Add
                    </button>
                </div>
            </div>
        </form>
    );
};
