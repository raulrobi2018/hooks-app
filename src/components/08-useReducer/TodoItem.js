import React from "react";

export const TodoItem = ({todo, index, handleDelete, handleToggle}) => {
    return (
        <div className="row mb-2">
            <div
                className={`${
                    todo.done
                        ? "col-10 text-left item complete"
                        : "col-10 text-left item"
                }`}
                onClick={() => handleToggle(todo.id)}
            >
                {index + 1} - {todo.desc}
            </div>
            <div className="col-2">
                <button
                    className="btn btn-danger"
                    onClick={() => handleDelete(todo.id)}
                    type="button"
                >
                    Delete
                </button>
            </div>
        </div>
    );
};
