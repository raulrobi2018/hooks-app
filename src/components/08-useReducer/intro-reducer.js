const initialState = [
    {
        id: 1,
        todo: "Buy bread",
        done: false
    }
];

// En un reducer, la acción es la que va a modificar el state
const todoReducer = (state = initialState, action) => {
    // action?: evalúa si action es undefined, si lo es, no lee la propiedad type
    if (action?.type === "add") {
        // No se debe hacer un push cuando estamos modificando el state,
        // Por eso aquí primero obtenemos todos los elementos del array con el operador spred (...state) y
        // luego le agregamos el que viene nuevo
        return [...state, action.payload];
    }

    return state;
};

let todoList = todoReducer();

const newTodo = {
    id: 2,
    todo: "Buy milk",
    done: false
};

const action = {
    type: "add",
    payload: newTodo
};

todoList = todoReducer(todoList, action);

console.log(todoList);
