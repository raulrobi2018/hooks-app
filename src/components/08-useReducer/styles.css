body {
    padding: 50px;
}

button {
    margin-left: 10px;
}