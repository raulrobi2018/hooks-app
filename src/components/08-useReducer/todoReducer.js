export const todoReducer = (state = [], action) => {
    switch (action.type) {
        case "add":
            // Retorna el state con el objeto agregado que viene en el payload
            return [...state, action.payload];
        case "delete":
            // Filtra el array de elementos del state y le borra el que pasamos en el payload
            return state.filter((todo) => todo.id !== action.payload);

        default:
            return state;
    }
};
