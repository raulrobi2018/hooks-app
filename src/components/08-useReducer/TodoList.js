import React from "react";
import { TodoItem } from "./TodoItem";

export const TodoList = (todoList, handleDelete, handleToggle) => {
  console.log(todoList);
  return (
    <>
      {todoList.map((t, i) => {
        return (
          <TodoItem
            todo={t}
            index={i}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
          />
        );
      })}
      ;
    </>
  );
};
