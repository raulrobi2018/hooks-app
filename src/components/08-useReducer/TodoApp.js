import React, {useEffect, useReducer} from "react";
import {useForm} from "../../hooks/useForm";

import "./styles.css";
import {todoReducer} from "./todoReducer";

// Inicializa el initialState
const init = () => {
    // Inicializo con los TODO's que estén guardados en el localstorage
    // Si no hay, el JSON.parse retorna 'null' y devolvemos un array vacío
    return JSON.parse(localStorage.getItem("todos")) || [];
};

export const TodoApp = () => {
    // La función init sirve para que el reducer no se esté ejecutando cada vez que se
    // renderiza el componente. En este caso lo que hacía el 'initialState' lo pasamos al init
    const [todoList, dispatch] = useReducer(todoReducer, [], init);

    // Utilizo mi custom hook definido en el archivo useForm.js
    // Extraigo la descripción del estado devuelto y la función que deseo manejar
    const [{description}, {handleInputChange, reset}] = useForm({
        description: ""
    });

    // Cuando la lista de TODO's tenga cualquier cambio; agrego, elimino o modifico en el localStorage
    useEffect(() => {
        localStorage.setItem("todos", JSON.stringify(todoList));
    }, [todoList]);

    const handleSubmit = (e) => {
        e.preventDefault();
        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        const action = {
            type: "add",
            payload: newTodo
        };

        // El dispatch será el encargado de cambiar el estado
        // con la información recibida en el payload
        dispatch(action);
        reset();
    };

    return (
        <>
            <h1>TODO Reducer</h1>
            <hr />
            <div className="row mb-3">
                <div className="col">
                    <strong>Number of TODO's: </strong> {todoList.length}
                </div>
            </div>

            <form className="mb-3">
                <div className="row">
                    <div className="col-8">
                        <input
                            type="text"
                            name="description"
                            placeholder="Description..."
                            autoComplete="off"
                            onChange={handleInputChange}
                            value={description}
                        ></input>
                    </div>
                    <div className="col-2">
                        <button
                            className="btn btn-success btn-block"
                            type="submit"
                            onClick={handleSubmit}
                        >
                            Add
                        </button>
                    </div>
                </div>
            </form>

            <hr />

            {todoList.map((t, i) => {
                return (
                    <div key={t.id} className="row mb-2">
                        <div className="col-10 text-left item">
                            {i + 1} - {t.desc}
                        </div>
                        <div className="col-2">
                            <button className="btn btn-danger">Delete</button>
                        </div>
                    </div>
                );
            })}
        </>
    );
};
