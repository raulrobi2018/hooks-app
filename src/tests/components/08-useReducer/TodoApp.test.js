import {mount, shallow, configure} from "enzyme";
import {TodoApp} from "../../../components/08-useReducer/TodoApp";
import {demoTodos} from "../../resources/demoTodos";
import {act} from "@testing-library/react";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";

configure({adapter: new Adapter()});

describe("Testing TodoApp component", () => {
    Storage.prototype.setItem = jest.fn(() => {});
    const wrapper = shallow(<TodoApp />);

    test("should return the component correctly", () => {
        expect(wrapper).toMatchSnapshot();
    });

    test("should add a Todo", () => {
        //El 'mount' funciona igual que el 'shallow' pero obtiene información
        //más completa acerca de sus hijos y otras cosas
        const wrapper = mount(<TodoApp />);

        act(() => {
            wrapper.find("TodoAdd").prop("addTodo")(demoTodos[0]);
            wrapper.find("TodoAdd").prop("addTodo")(demoTodos[1]);
        });

        expect(wrapper.find("#numElements").text().trim()).toBe(
            "Number of TODO's:  2"
        );

        //2 veces porque lo ejecutamos 2 veces en el act
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    });

    test("should delete a todo", () => {
        wrapper.find("TodoAdd").prop("addTodo")(demoTodos[0]);
        wrapper.find("TodoList").prop("handleDelete")(demoTodos[0].id);

        expect(wrapper.find("#numElements").text().trim()).toBe(
            "Number of TODO's:  0"
        );
    });
});
