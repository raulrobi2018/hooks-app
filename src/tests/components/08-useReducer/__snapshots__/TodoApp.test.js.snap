// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing TodoApp component should return the component correctly 1`] = `
<Fragment>
  <h1>
    TODO Reducer
  </h1>
  <hr />
  <div
    className="row mb-3"
  >
    <div
      className="col"
      id="numElements"
    >
      <strong>
        Number of TODO's: 
      </strong>
       
      0
    </div>
  </div>
  <TodoAdd
    addTodo={[Function]}
  />
  <hr />
  <TodoList
    handleDelete={[Function]}
    handleToggle={[Function]}
    todoList={Array []}
  />
</Fragment>
`;
